src/sa.py

```diff
@@ -11,5 +17,103 @@ def main():
           f"in genome {args.genome.name}")
 
 
+def sa_construction_nsq(x: str) -> list[int]:
+    '''
+    Suffix array construction in O(n^2), using radix sort
+
+
+    Put input string into a set O(n) -> sort set to make alphabet string O(n * alphabet)
+     -alphabet size is bounded by n anyway.
+    Buckets constructed as dict, in order of alphabet string. O(n)
+    Input string is split into suffixes, adding $ so each has length n, O(n*n) 
+     -could get rid of slicing and just use index, but am lazy, and already implemented radix using strings (and we're n^2 anyway)
+    radix sort suffixes O(n*n)
+    one last O(n) running through sorted suffixes to get their starting index (because lazyness.)
+    '''
+    n = len(x)
+    alpha = alphabet_from_input_string(x)
+    suffixes = [x[i:]+i*'$' for i, _ in enumerate(x)]
+
+    sort = radix_sort(suffixes, alpha)
+    return [n-len(x.split('$')[0])-1 for x in sort]
+
+
+def sa_construction_simple_nsq(x: str) -> list[int]:
+    '''
+    simple sa construction uing builtin strcmp and sort, for comparisson
+    '''
+    suf = [x[i:] for i, _ in enumerate(x)]
+    suf = sorted(suf)
+    return [len(x) - len(i) for i in suf]
+
+
+def lower_bound(a: str, i: int, lo: int, hi: int, x: str, sa: list[int]) -> int:
+    ''' binary search suffix array for lower bound for occurence of character a at position i in a pattern, in string x, given Suffix array SA of x'''
+    while hi > lo:
+        mid = (hi+lo)//2
+        if x[(sa[mid] + i) % len(x)] < a:
+            lo = mid+1
+        else:
+            hi = mid
+    return lo
+
+
+def upper_bound(a: str, i: int, lo: int, hi: int, x: str, sa: list[int]) -> int:
+    ''' binary search suffix array for upper bound for occurence of character a at position i in a pattern, in string x, given Suffix array SA of x'''
+    while hi > lo:
+        mid = (hi+lo)//2
+        if ord(x[(sa[mid] + i) % len(x)]) < ord(a)+1:
+            lo = mid+1
+        else:
+            hi = mid
+    return lo
```

Comment:
I think `return lower_bound(chr(ord(a)+1), ...)` might be a shorter implementation...